### PR TITLE
[Flink AI Flow] [Document] Add tutorial document

### PR DESCRIPTION
**Pull request by bgeng777**
## What is the purpose of the change

Add tutorial document

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

### CODE REVIEW

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+# scheduler config
+scheduler:
+  scheduler_class_name: ai_flow_plugins.scheduler_plugins.airflow.airflow_scheduler.AirFlowScheduler
+  scheduler_config:
+    airflow_deploy_path: /absolute/path/to/airflow/airflow_deploy
+    notification_service_uri: localhost:50052
+```
+
+## Prepare Project Directory
+Before writing the workflow, we should prepare a project directory whose structure is as follows:
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:45]: prepare a project directory structured as follows

Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:45]: prepare a project directory structured as follows

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+server_ip: localhost
+server_port: 50051
+blob:
+  blob_manager_class: ai_flow_plugins.blob_manager_plugins.local_blob_manager.LocalBlobManager
+```
+In `project_name`, we define the project name, which will be the default namespace of workflows in this project as well. 
+
+> Namespace in Flink AI Flow is used for isolation. Each workflow can only send events to its own namespace while it can listen on multiple namespaces. The reason for enabling listening on multiple namespaces is that the workflow could be triggered by external events from notification service.
+
+For `server_ip` and `server_port`,  they tell where the `AIFlowServer` will starts on.
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:83]: they tell where the `AIFlowServer` is running on.

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+Then, we configure the `blob` property which tells Flink AI Flow which manages the execution environment(i.e., local or remote env) of workflows in this project.
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:85]: tells Flink AI Flow about what?

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+
+Here we choose to use `LocalBlobManager` and as a result, the project will run locally. 
+
+We currently also provide `OssBlobManager` which allows users to run their workflow remotely.
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:89]: submit and run?

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+
+### tutorial_workflow.yaml 
+Next, we will introduce how to write the workflow configuration yaml file.
+```yaml
+train:
+  job_type: python
+  periodic_config:
+    interval: '0,0,0,60' # The train job will be executed 60s
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:97]: will be executed every 60s

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+For `validate` job, we only config its job type to be `python`.
+
+For `predict` job, we set its job type to be `flink`, which means this job is a flink job. In addition, to use flink, we need to set some flink-specific properties including `run_mode` (cluster or local) and `flink_run_args`. For the `flink_run_args`, we follow the list format of `yaml` to add the args such as the `-pyexec`.
+
+## Writing the Workflow Code
+
+Now, let's start coding to implement workflow described in Target section
+
+### Importing Modules
+A workflow Flink AI Flow is just a Python script. Let’s create `tutorial_workflow.py` and start by importing the libraries we will need.
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:123]: A workflow in Flink AI Flow

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+from ai_flow.util.path_util import get_file_dir
+
+# User-defined processors; we will introduce them in tutorial_processors.py
+from tutorial_processors import DatasetReader, ModelTrainer, ValidateDatasetReader, ModelValidator, Source, Sink, \
+    Predictor
+
+```
+
+###  Defining a Training Job
+In our design, the workflow in Flink AI Flow is a DAG(Directed Acyclic Graph) or to be more specifically, it is a [AIGraph](). Each node in the graph is an [AINode](), which contains a processor. Users should write their custom logic in the processor.
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:145]: to be more specific

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+These concepts seem to be a little difficult, but do not worry. Let's look at some code snippets, and you will find it pretty easy to define a job with provided API of Flink AI Flow.
+
+In the following codes, we define a training job of our workflow:
+```python
+    artifact_prefix = af.current_project_config().get_project_name() + "."
+    # Training of model
+    with af.job_config('train'):
+        # Register metadata of training data(dataset), `uri` refers to the file path
+        train_dataset = af.register_dataset(name=artifact_prefix + 'train_dataset',
+        																		uri=DATASET_URI.format('train'))
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:162]: where is DATASET_URI defined?

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+
+In Flink AI Flow, `Channel` represents the output of AINodes. More vividly, `Channel` is one end of the `DataEdge`. The following picture shows the relation among `AINodes`, `DataEdges` and `Channels`:
+
+![](images/tutorial/channels.png)
+
+In the example, AINode N0 has 3 outputs(i.e., 3 channels whose source node is N0). We can make N1 accepts the c0 and c1 channels and N2 accepts c1 and c2 channels. Accordingly, there are 4 DataEdges in all. With such design, we can manipulate the data dependencies more flexibly and reuse the data easier.
+
+
+
+Currently, the only puzzle left is how do we implement the processors including `DatasetReader()` and `ModelTrainer()`. We will introduce them later in [Implementing custom processor Processors](#Implementing custom processors) section. Next, let's pay attention to defining the Validation and Prediction jobs.
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:188]: the link doesn't work?

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+
+Above codes show the basic use of workflow administration API. You may notice that our naming convention of workflow management API seems inconsistent: `submit_workflow()` and `start_new_workflow_execution()`. 
+
+In fact, in our design, `Workflow` defines the execution logic of a set of jobs together with control edges among them. But after definition, the workflow will not be run. Instead, the workflow will be scheduled and triggered by the scheduler(e.g., Airflow scheduler) after calling `start_new_workflow_execution()`. The execution entity of workflow is named as `WorkflowExecution`, which is created by the scheduler. One advantage of this design is that reproducing a workflow can be much easier as we can locate the past workflow execution with all environments conveniently.
+
+
+
+## Implementing custom processors
+
+As we have mentioned, users need to write their own logic in processors for each job. Currently, Flink AI Flow supports `baash`, `python` and `flink` processors.
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:299]: typo: bash

File: flink-ai-flow/wiki_doc/Tutorial.md [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+class DatasetReader(PythonProcessor):
+
+    def process(self, execution_context: ExecutionContext, input_list: List) -> List:
+        """
+        Read dataset using pandas
+        """
+        # Gets the registered dataset meta info
+        dataset_meta: af.DatasetMeta = execution_context.config.get('dataset')
+        # Read the file using pandas
+        train_data = pd.read_csv(dataset_meta.uri, header=0, names=EXAMPLE_COLUMNS)
```
Comment (Sxnan) [inline flink-ai-flow/wiki_doc/Tutorial.md:313]: EXAMPLE_COLUMNS is not defined

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 1 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:45]: fixed

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 2 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:83]: fixed

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 3 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:85]: fixed

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 4 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:89]: fixed

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 5 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:97]: fixed

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 6 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:123]: In Flink AI Flow, a workflow is just a Python script.

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 7 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:145]: fixed

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 8 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:162]: Add DATASET_URI in ### Importing Modules section.

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 10 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:299]: fixed

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 11 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md:313]: Added.

File: flink-ai-flow/wiki_doc/Tutorial.md (hunk 9 above)
Comment (bgeng777) [reply flink-ai-flow/wiki_doc/Tutorial.md]: fixed

**Review by Sxnan — APPROVED** [review]
